File: sdk-csharp/DepHealth.Core/EndpointState.cs
namespace DepHealth;

/// <summary>
/// Thread-safe endpoint state: healthy/unhealthy, counters for consecutive successes/failures,
/// and detailed status for the HealthDetails() API.
/// </summary>
internal sealed class EndpointState
{
    private readonly object _lock = new();
    private bool? _healthy;       // null = UNKNOWN
    private int _consecutiveFailures;
    private int _consecutiveSuccesses;

    // Dynamic fields (updated under lock by StoreCheckResult)
    private string _lastStatus = StatusCategory.Unknown;
    private string _lastDetail = "unknown";
    private TimeSpan _lastLatency;
    private DateTimeOffset? _lastCheckedAt;

    // Static fields (set once before Start via SetStaticFields)
    private string _depName = "";
    private string _depType = "";
    private string _host = "";
    private string _port = "";
    private bool _critical;
    private Dictionary<string, string> _labels = new();

    public bool? Healthy
    {
        get { lock (_lock) { return _healthy; } }
    }

    public void SetStaticFields(
        string depName, string depType, string host, string port,
        bool critical, IReadOnlyDictionary<string, string> labels)
    {
        _depName = depName;
        _depType = depType;
        _host = host;
        _port = port;
        _critical = critical;
        _labels = new Dictionary<string, string>(labels);
    }

    public void StoreCheckResult(string category, string detail, TimeSpan latency)
    {
        lock (_lock)
        {
            _lastStatus = category;
            _lastDetail = detail;
            _lastLatency = latency;
            _lastCheckedAt = DateTimeOffset.UtcNow;
        }
    }

    public EndpointStatus ToEndpointStatus()
    {
        lock (_lock)
        {
            return new EndpointStatus(
                healthy: _healthy,
                status: _lastStatus,
                detail: _lastDetail,
                latency: _lastLatency,
                type: _depType,
                name: _depName,
                host: _host,
                port: _port,
                critical: _critical,
                lastCheckedAt: _lastCheckedAt,
                labels: new Dictionary<string, string>(_labels));
        }
    }

    public void RecordSuccess(int successThreshold)
    {
        lock (_lock)
        {
            _consecutiveFailures = 0;
            _consecutiveSuccesses++;

            if (_healthy is null)
            {
                // First check — immediate transition
                _healthy = true;
                return;
            }

            if (!_healthy.Value && _consecutiveSuccesses >= successThreshold)
            {
                _healthy = true;
            }
        }
    }

    public void RecordFailure(int failureThreshold)
    {
        lock (_lock)
        {
            _consecutiveSuccesses = 0;
            _consecutiveFailures++;

            if (_healthy is null)
            {
                // First check — immediate transition
                _healthy = false;
                return;
            }

            if (_healthy.Value && _consecutiveFailures >= failureThreshold)
            {
                _healthy = false;
            }
        }
    }
}

File: sdk-java/dephealth-core/src/main/java/biz/kryukov/dev/dephealth/scheduler/EndpointState.java
package biz.kryukov.dev.dephealth.scheduler;

import biz.kryukov.dev.dephealth.DependencyType;
import biz.kryukov.dev.dephealth.EndpointStatus;
import biz.kryukov.dev.dephealth.StatusCategory;

import java.time.Duration;
import java.time.Instant;
import java.util.Map;
import java.util.concurrent.ScheduledFuture;

/**
 * Thread-safe endpoint state: healthy/unhealthy, consecutive success/failure counters,
 * and fields for the HealthDetails() API.
 */
public final class EndpointState {

    private Boolean healthy;         // null = UNKNOWN
    private int consecutiveFailures;
    private int consecutiveSuccesses;

    // Dynamic fields for HealthDetails() API.
    private String lastStatus;
    private String lastDetail;
    private Duration lastLatency;
    private Instant lastCheckedAt;

    // Scheduled future for cancellation (used by dynamic endpoint management).
    private ScheduledFuture<?> future;

    // Static fields set at state creation time.
    private String depName;
    private DependencyType depType;
    private String host;
    private String port;
    private boolean critical;
    private Map<String, String> labels;

    public synchronized Boolean healthy() {
        return healthy;
    }

    public synchronized void recordSuccess(int successThreshold) {
        consecutiveFailures = 0;
        consecutiveSuccesses++;

        if (healthy == null) {
            // First check — immediate transition
            healthy = true;
            return;
        }

        if (!healthy && consecutiveSuccesses >= successThreshold) {
            healthy = true;
        }
    }

    public synchronized void recordFailure(int failureThreshold) {
        consecutiveSuccesses = 0;
        consecutiveFailures++;

        if (healthy == null) {
            // First check — immediate transition
            healthy = false;
            return;
        }

        if (healthy && consecutiveFailures >= failureThreshold) {
            healthy = false;
        }
    }

    /**
     * Sets the static fields for this endpoint (called once during registration).
     */
    synchronized void setStaticFields(String depName, DependencyType depType, String host,
                                       String port, boolean critical,
                                       Map<String, String> labels) {
        this.depName = depName;
        this.depType = depType;
        this.host = host;
        this.port = port;
        this.critical = critical;
        this.labels = labels == null ? Map.of() : Map.copyOf(labels);
        // Set UNKNOWN defaults for dynamic fields.
        this.lastStatus = StatusCategory.UNKNOWN;
        this.lastDetail = StatusCategory.UNKNOWN;
        this.lastLatency = Duration.ZERO;
        this.lastCheckedAt = null;
    }

    /**
     * Stores the classification results from a health check.
     */
    synchronized void storeCheckResult(String status, String detail, Duration latency) {
        this.lastStatus = status;
        this.lastDetail = detail;
        this.lastLatency = latency;
        this.lastCheckedAt = Instant.now();
    }

    /**
     * Returns the scheduled future for this endpoint's periodic check.
     */
    synchronized ScheduledFuture<?> future() {
        return future;
    }

    /**
     * Sets the scheduled future for this endpoint's periodic check.
     */
    synchronized void setFuture(ScheduledFuture<?> future) {
        this.future = future;
    }

    /**
     * Creates an {@link EndpointStatus} snapshot of the current state.
     */
    synchronized EndpointStatus toEndpointStatus() {
        return new EndpointStatus(
                healthy,
                lastStatus,
                lastDetail,
                lastLatency,
                depType,
                depName,
                host,
                port,
                critical,
                lastCheckedAt,
                labels
        );
    }
}
